pkg/auth/oauth/dynamic_registration.go

```diff
@@ -0,0 +1,240 @@
+		Scopes:                  scopes,
+	}
+	return registrationRequest
+}
+
+// DynamicClientRegistrationResponse represents the response from dynamic client registration (RFC 7591)
+type DynamicClientRegistrationResponse struct {
+	// Required fields
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret,omitempty"`
+
+	// Optional fields that may be returned
+	ClientIDIssuedAt        int64  `json:"client_id_issued_at,omitempty"`
+	ClientSecretExpiresAt   int64  `json:"client_secret_expires_at,omitempty"`
+	RegistrationAccessToken string `json:"registration_access_token,omitempty"`
+	RegistrationClientURI   string `json:"registration_client_uri,omitempty"`
+
+	// Echo back the essential request fields
+	ClientName              string   `json:"client_name,omitempty"`
+	RedirectURIs            []string `json:"redirect_uris,omitempty"`
+	TokenEndpointAuthMethod string   `json:"token_endpoint_auth_method,omitempty"`
+	GrantTypes              []string `json:"grant_types,omitempty"`
+	ResponseTypes           []string `json:"response_types,omitempty"`
+	Scopes                  []string `json:"scope,omitempty"`
+}
+
+// RegisterClientDynamically performs dynamic client registration (RFC 7591)
+func RegisterClientDynamically(
+	ctx context.Context,
+	registrationEndpoint string,
+	request *DynamicClientRegistrationRequest,
+) (*DynamicClientRegistrationResponse, error) {
+	return registerClientDynamicallyWithClient(ctx, registrationEndpoint, request, nil)
+}
+
+// validateRegistrationEndpoint validates the registration endpoint URL
+func validateRegistrationEndpoint(registrationEndpoint string) (*url.URL, error) {
+	registrationURL, err := url.Parse(registrationEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("invalid registration endpoint URL: %w", err)
+	}
+
+	// Ensure HTTPS for security (except localhost for development)
+	if registrationURL.Scheme != "https" && !networking.IsLocalhost(registrationURL.Host) {
+		return nil, fmt.Errorf("registration endpoint must use HTTPS: %s", registrationEndpoint)
+	}
+
+	return registrationURL, nil
+}
+
+// validateAndSetDefaults validates the request and sets default values
+func validateAndSetDefaults(request *DynamicClientRegistrationRequest) error {
+	if request == nil {
+		return fmt.Errorf("registration request cannot be nil")
+	}
+	if len(request.RedirectURIs) == 0 {
+		return fmt.Errorf("at least one redirect URI is required")
+	}
+
+	// Set default values if not provided
+	if request.ClientName == "" {
+		request.ClientName = ToolHiveMCPClientName
+	}
+	if len(request.GrantTypes) == 0 {
+		request.GrantTypes = []string{"authorization_code"}
+	}
+	if len(request.ResponseTypes) == 0 {
+		request.ResponseTypes = []string{"code"}
+	}
+	if request.TokenEndpointAuthMethod == "" {
+		request.TokenEndpointAuthMethod = "none" // For PKCE flow
+	}
+
+	return nil
+}
+
+// createHTTPRequest creates the HTTP request for dynamic client registration
```

Comment:
could we isolate the http creation/management into a different utils file? i believe this file is getting very large